### DISCUSSION TITLE
Creating a validation wrapper extending valibot schemas with additional actions

### DISCUSSION THREAD

**Discussion by imolorhe** [open-circle/valibot#898] [Q&A]
Hey! 👋🏾 I'm working on a project that uses valibot for validating input and output data based on a schema. I'm able to specify the type as `GenericSchema` for the schema which allows any valibot schema to be used. However for my project, I want to add additional custom checks/actions to the regular valibot schemas. For example, validating that strings don't contain a specific substring, or numbers follow an expected pattern, etc. Even for schemas that don't need any additional checks (yet), I want to create a wrapping schema for them and require the use of the project schemas instead of the regular valibot schemas. That is, using schemas (`object()`, `string()`, `number()`, etc) from valibot should show a typescript error and a runtime error and require the use of equivalent schemas with the additional validations from the project instead.

Is there a way to achieve this without having to re-implement the schemas and parsers already implemented in valibot?

**Comment by imolorhe** [comment]
Hmm, maybe [brand](https://valibot.dev/api/brand/) is what I'm looking for?

**Reply by imolorhe** [reply]
Okay looks like that just affects the output type.

**Comment by fabian-hiller** [comment]
You can extend the validation of an existing schema using our `pipe` method:

```ts
import * as v from 'valibot';

const EmailSchema = v.pipe(v.string(), v.email());
const GmailSchema = v.pipe(EmailSchema, v.endsWith('@gmail.com'));
```

And here is an example that shows how you can only allow specific schemas as arguments when using TypeScript:

```ts
import * as v from 'valibot';

const EmailSchema = v.pipe(v.string(), v.email());
type  EmailSchema = typeof EmailSchema;

const LoginSchema = v.object({
  email: EmailSchema,
  password: v.pipe(v.string(), v.minLength(8)),
});
type LoginSchema = typeof LoginSchema;

function foo(schema: EmailSchema | LoginSchema) {
  console.log(schema);
}

foo(EmailSchema); // Works
foo(v.string()); // Error
```

**Reply by imolorhe** [reply]
Thanks! This is slightly different from what I wanted though. While the above works, I don't want to have specific schemas but instead to allow the combination of the different schemas and actions in the same way you can with the direct valibot schemas, but somehow typing the wrapper (custom) validators. Just noticed that might be an issue though since it would lead to nested pipe methods (i.e. pipe used in the custom schema and the user wanting to add additional methods/actions) which I don't think is supported.

**Reply by fabian-hiller** [reply]
Can you provide example code? The code don't have to work. I just need more info to be able to tell you if it is supported.

**Reply by imolorhe** [reply]
Sure. Here's a sample code

```js
import {string, object, email, pipe} from 'src/shared/validator';
import * as v from 'valibot';

//✅ Schema generated from wrapper validators (should work)
const MySchema = object({
  first: string(),
  email: pipe(string(), email())
});

// ❌ Schema directly using valibot validators, meaning the additional checks can't be performed (shouldn't work)
const MySchema2 = v.object({
  first: v.string(),
  email: v.pipe(v.string(), v.email())
});

// .. in src/shared/validator
// Should still be composable
export const string = (...args) => {
  return v.pipe(string(...args), v.check(input => !input.includes('secret_value')));
}
export const pipe = (...args) => {
  return v.pipe;
}
// ...etc
```

While underneath the validator uses valibot schemas and parser, I wouldn't want users of the project to "mistakenly" call valibot validators directly since those will not have the additional checks we want to apply, but at the same time, I would want the users to be able to compose their schema using `pipe` as well, which wouldn't work here since we will already be using the pipe to add the additional validation checks

**Reply by fabian-hiller** [reply]
I think what you have in mind, should work. But it requires a lot of work on your side to make it perfect.

**Reply by imolorhe** [reply]
Got it. So the other approach below is a more _supported_ approach then

**Reply by fabian-hiller** [reply]
I would say that it depends on the use case.

**Reply by imolorhe** [reply]
Fair enough, but would like to verify that the approach of _modifying_ schemas is fine and I'm not missing some other parts of the schema that need to be updated

**Reply by fabian-hiller** [reply]
People rarely modify the schema object, but we do it internally, for example with `pick` and `omit`. So Valibot is written with this in mind.

**Reply by imolorhe** [reply]
Thanks! Sounds good to me!

**Reply by fabian-hiller** [reply]
It's more likely to make mistakes if you do, and we don't expect external people to do it either. That's why every schema property is marked read-only, but in practice it should work in most cases. So if you know what you are doing it should be fine.

**Comment by imolorhe** [comment]
@fabian-hiller For my use case, I decided to go a different route where I _modify_ the schema based on the value in the metadata action I find. This removes the need for context based runtime validation, and instead I want to modify the schema based on the context, which should mean better performance (since the additional context-based logic isn't executed at runtime). I am following a similar pattern used in [getDefaults](https://valibot.dev/api/getDefaults/) and [getFallbacks](https://valibot.dev/api/getFallbacks/) to iterate over all the schemas in the schema. When I find a metadata value, I decide to either keep the schema or delete it from the schema object.

This appears to be working, when I parse some data using the modified schema, the dropped fields don't get returned. However I'm not sure if this is a supported pattern that could lead to unexpected behaviors, or if there's a different approach I should be taking here. For example, something I would like to do is add additional actions as well which I think I can achieve by creating a `schema.pipe` list (if it doesn't already exist) with the additional actions, but it feels like I'm working off of internal APIs

**Reply by fabian-hiller** [reply]
If you correctly work with our external API this is totally fine. Feel free to provide us feedback if there is anything we can do better.

**Reply by imolorhe** [reply]
Here's an example of what I have

```js

export function getModifiedSchema(schema) {
  if ('pipe' in schema) {
    const metadataAction = schema.pipe.find(item => item.type === 'metadata');
    
    if (metadataAction?.metadata.allowed) {
      return schema;
    }
    
    if (metadataAction?.metadata.checkSecret) {
      // haven't tested this one, but this is something else I'd likely be doing..
      schema.pipe.push(v.check(
        (input) => typeof input === 'string' && !input.includes('secret_value'),
        'The item cannot have the secret value'
      ));
      return schema;
    }
  }
  
  return undefined;
}

export function getModifiedSchemas(schema) {
  // If it is an object schema, return defaults of entries
  if ('entries' in schema) {
    Object.entries(schema.entries).forEach(([key, value]) => {
      const modifiedSchema = getModifiedSchemas(value);
      if (!modifiedSchema) {
        Reflect.deleteProperty(schema.entries, key);
      }
    });
    return schema;
  }

  // If it is a tuple schema, return defaults of items
  if ('items' in schema) {
    return schema.items.map(getDefaults).filter(Boolean);
  }

  return getModifiedSchema(schema);
}

```

The schema can either be deleted or modified
